lib/elixir/lib/calendar/duration.ex

```diff
@@ -331,4 +333,100 @@ defmodule Duration do
         raise ArgumentError, ~s/failed to parse duration "#{string}". reason: #{inspect(reason)}/
     end
   end
+
+  @doc """
+  Converts the given duration to [ISO 8601-2:2019](https://en.wikipedia.org/wiki/ISO_8601) format.
+
+  Note this function implements the *extension* of ISO 8601:2019. This extensions allows weeks to
+  appear in any case, which is not the case in the base standard. `"P3M3W"` is not allowed in
+  ISO 8601-1, but it is in ISO 8601-2. Thus, any duration can be output by this function.
+
+  ## Examples
+
+      iex> Duration.new!(year: 3) |> Duration.to_iso8601()
+      "P3Y"
+      iex> Duration.new!(day: 40, hour: 12, minute: 42, second: 12) |> Duration.to_iso8601()
+      "P40DT12H42M12S"
+      iex> Duration.new!(second: 30) |> Duration.to_iso8601()
+      "PT30S"
+      iex> Duration.new!(second: 30) |> Duration.multiply(-1) |> Duration.to_iso8601()
+      "PT-30S"
+
+      iex> Duration.new!([]) |> Duration.to_iso8601()
+      "PT0S"
+
+      iex> Duration.new!(second: 1, microsecond: {2_200, 3}) |> Duration.to_iso8601()
+      "PT1.002S"
+      iex> %Duration{second: 1, microsecond: {-1_200_000, 4}} |> Duration.to_iso8601()
+      "PT-0.2000S"
+  """
+
+  @spec to_iso8601(t) :: String.t()
+  def to_iso8601(duration)
+
+  def to_iso8601(%Duration{
+        year: 0,
+        month: 0,
+        week: 0,
+        day: 0,
+        hour: 0,
+        minute: 0,
+        second: 0,
+        microsecond: {0, _}
+      }) do
+    "PT0S"
+  end
+
+  def to_iso8601(%Duration{} = d) do
+    IO.iodata_to_binary([?P, to_iso8601_duration_date(d), to_iso8601_duration_time(d)])
+  end
```

Comment:
Given the period prefix `-P10D` might be more common than the individual prefix `P-10D` (since it's actually specified in the ISO 8601-2 extension) I wonder if we want to check if _all_ components are negative, so that we convert the duration like so:

```
iex> Duration.to_iso8601(%Duration{day: -10, hour: -2})
"-P10DT2H"

iex> Duration.to_iso8601(%Duration{day: 10, hour: -2})
"P10DT-2H"
```

Just a suggestion since we anyways support both now. i am also fine with always just negating the negative components for the sake of performance.

Reply:
> with seconds and microseconds, it's not that simple: we need to check the total of microseconds is negative ... and take into account precision

Right, for the first part, here is a bad idea: we could use `to_timeout/1` on the `duration` to check if the duration is overall negative or not. To your second point, I dont think converting `%Duration{microsecond: {-2, 3}}` to `-PT0S` would be wrong, however i'd also prefer we didn't 😅 

> some ISO8601-1 libraries can't parse it

As far as i've seen this would be especially true for negative duration units given they aren't actually specified, while the leading minus sign is at least specified in ISO 8601-2.

Definitely just a suggestion which some obvious downsides - let's see if others have an opinion on it, i am certainly not pushing for either direction as well.